src/ds/ptrwrap.h

```diff
@@ -0,0 +1,142 @@
+#pragma once
+
+namespace snmalloc
+{
+  /**
+   * A convenience type that we can use to annotate internal functions
+   * which return pointers that are headed out to the application or
+   * which have come back.
+   */
+  class ReturnPtr
+  {
+  public:
+    void* unsafe_return_ptr;
+
+    ReturnPtr(const std::nullptr_t n)
```

Comment:
Don't we need a constructor from `AuthPtr`?

Reply:
No, by design: we're trying to ensure that we go through `ptrauth_bound`, which takes an `AuthPtr` to a `FreePtr`, and then `unsafe_mk_returnptr` to convert the `FreePtr` to `ReturnPtr`.  The intention is that most allocation paths actually only need the latter half, as they will have obtained a `FreePtr` directly from a fast free list or such, using any `AuthPtr`s they have just to get to allocator metadata (e.g., using `Superslab::get(AuthPtr)` but then reading out a `FreePtr`).

Reply:
It feels like the idiomatic way of doing that would be to remove `unsafe_mk_freeptr`, add a private constructor to `FreePtr` that takes an `AuthPtr` and make the few functions that should call it `friend`s of `FreePtr`.